ddp-workspace/projects/ddp-sdk/src/lib/services/activity/validators/activityUniqueValidationRule.ts

```diff
@@ -0,0 +1,34 @@
+import { ActivityQuestionBlock } from '../../../models/activity/activityQuestionBlock';
+import { ActivityAbstractValidationRule } from './activityAbstractValidationRule';
+import { QuestionType } from '../../../models/activity/questionType';
+import { ActivityCompositeQuestionBlock, ActivityPicklistQuestionBlock } from 'ddp-sdk';
+import { PicklistSelectMode } from '../../../models/activity/picklistSelectMode';
+
+export class ActivityUniqueValidationRule extends ActivityAbstractValidationRule {
+    constructor(question: ActivityQuestionBlock<any>) {
+        super(question);
+    }
+
+    public recalculate(): boolean {
+        const {answer: answers} = this.question;
+        if (this.question.questionType !== QuestionType.Composite || answers.length === 1) {
+            this.result = null;
+            return true;
+        }
+
+        const [firstChild] = (this.question as ActivityCompositeQuestionBlock).children;
+
+        if (firstChild.questionType !== QuestionType.Picklist
+            || (firstChild as ActivityPicklistQuestionBlock).selectMode !== PicklistSelectMode.SINGLE) {
+            this.result = null;
+            return true;
```

Comment:
I was expecting that our implementation could deal with with cases beyond picklist.
e.g., so if we have a composite that the child is just a text question,I would like to use the UNIQUE rule to flag that somebody typed "FOO" twice/in two children.

Reply:
To support all the different non-composite question comparisons, it might make sense to factor out to the individual answers or blocks the comparison logic, e.g., add an abstract method like `isEqual(other: T)` to ActivityQuestionBlock . Leave it unimplemented in abstract class.
This makes it easier to keep Unique rule up to date over time/as we add more question types; if a new question type is added, you are forced to create a new question block sub type. And then the compiler remind you that you have to implements isEqual. Otherwise we always have to remember that UNIQUE rule code needs to be updated whenever a new question type comes along. I don't want to have to remember that.

Reply:
totally agree, done!

Reply:
If I am reading your solution, I think it works if we have the one child question (like in cancer/single picklist) but I think it won't work if a composite is made up of multiple questions, e.g., a date and a number.

I think the use of Sets to find duplicates is generally a good idea though I personally am disappointed that out-of-the box implementation does not allow you in Javascript to define your one equality/depends on the == operator (maybe there is a way to override prototype implementation?

It occurred to me that you can PROBABLY get away with a single implementation of compare regardless of the answer type. Would look something like this in the AbstractActivityQuestionBlock (using the Underscore isEqual implementation)

```typescript
public answersAreEqual(answer1: T, answer2: T): boolean {
        return _.isEqual(answer1, answer2);
    }
```
Can potentially be overridden if for some reason we thought for a particular answer type` _.isEqual` would not work

So then trick is to go through the composite answers and see if something in one of the answer columns is different. Looks something like this (not tested! just for discussion purposes)

```typescript
const childQuestions: ActivityQuestionBlock<any>[] = (this.question as ActivityCompositeQuestionBlock).children;
const rowsOfAnswers: AnswerContainer[][] = (this.question as ActivityCompositeQuestionBlock).answer;
        childQuestions.forEach((currentChildQuestion, questionIdx) => {
            let valueInFirstRow: any;
            rowsOfAnswers.forEach((rowOfAnswers: AnswerContainer[], answerRowIdx) => {
                const currentAnswerValue = rowOfAnswers[questionIdx].value;
                if (answerRowIdx === 0) {
                    valueInFirstRow = currentAnswerValue;
                } else if (!currentChildQuestion.answersAreEqual(valueInFirstRow, currentAnswerValue)) {
                    return false;
                }
            });
        });
        return true;
```

And if we confident that `_isEqual` always works for all our answers, then we could just use it directly in the rule and not bother implementing answersAreEqual() .

At any rate, just illustration of what was in my head. But just make sure that your implementation works with composites that have multiple child questions.

Reply:
1. Is it possible composite question has questions of multiple types? like number and date? I thought the implementation allows us to add only one type on question to composite question, no?
2. In general if we want to compare 2 different types it's always a custom logic. So I'm not sure what's your suggestion here. Do you want to have implementation to compare each type of question with each type of questions? it sounds like plenty of logic and if there is no the case but just thoughs how to make it more abstract I'd suggest to keep for now the implementation of comparing the answers of only one type.

Reply:
> 1. Is it possible composite question has questions of multiple types? like number and date? I thought the implementation allows us to add only one type on question to composite question, no?
Composites questions are designed to have one or more child questions. In the case of the cancer question, it is one child picklist question.
But here is one with a text question and a date as children:
![Screen Shot 2021-09-02 at 12 25 27 PM](https://user-images.githubusercontent.com/29984380/131881609-8d4d7a2d-4e16-4afa-b425-232f6f71db08.png)
The allowMultiple=true|false determines whether we can have multiple rows of answers. 


> 2. In general if we want to compare 2 different types it's always a custom logic. So I'm not sure what's your suggestion here. Do you want to have implementation to compare each type of question with each type of questions? it sounds like plenty of logic and if there is no the case but just thoughs how to make it more abstract I'd suggest to keep for now the implementation of comparing the answers of only one type.
Well, custom logic might not be necessary.
For the composite we DO KNOW what are the child question types. And each question type has a specific answer type, e.g.:

Reply:
ooohhh I see now, thanks a lot for that example with multiple question blocks per one composite question! I didn't thought about this case actually... But it's fixed now, please take a look. I hope I took into account all cases :) but feel free to add more comments if you feel something is missing